src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp

```diff
@@ -0,0 +1,191 @@
+      CHECK(stop_called_);
+    }
+
+    CHECK(data_produce_thread_.joinable());
+    data_produce_thread_.join();
+
+    CHECK(data_upload_thread_.joinable());
+    data_upload_thread_.join();
+  }
+
+  UploaderClientCaptureEventCollector collector_;
+  size_t total_uploaded_data_bytes_ = 0;
+
+  absl::Mutex upload_finished_mutex_;
+  bool upload_finished_ ABSL_GUARDED_BY(upload_finished_mutex_) = false;
+  absl::Mutex early_stop_mutex_;
+  bool early_stop_ ABSL_GUARDED_BY(early_stop_mutex_) = false;
+
+ private:
+  // Keep producing capture events until stop is called or `event_count_` events are produced with a
+  // specified produce duration. The last produced event will always be the `CaptureFinishedEvent`.
+  void ProduceCaptureData() {
+    size_t added_event_count = 0;
+    bool need_check_early_stop = (early_stop_event_count_ >= 2);
+    bool stopped = false;
+    while (added_event_count < event_count_ - 1 && !stopped) {
+      if (need_check_early_stop && added_event_count + 2 >= early_stop_event_count_) {
+        {
+          absl::MutexLock lock{&early_stop_mutex_};
+          early_stop_ = true;
+        }
+        need_check_early_stop = false;
+      }
+
+      stop_called_mutex_.LockWhenWithTimeout(absl::Condition(&stop_called_),
+                                             absl::Milliseconds(produce_event_every_ms_));
+      collector_.AddEvent(CreateInternedStringCaptureEvent(kAnswerKey, kAnswerString));
+      stopped = stop_called_;
+      stop_called_mutex_.Unlock();
+
+      ++added_event_count;
+    }
+
+    collector_.AddEvent(CreateCaptureFinishedEvent());
+
+    Stop();
+  }
+
+  void UploadCaptureData() {
+    while (collector_.GetDataReadiness() != orbit_capture_uploader::DataReadiness::kEndOfData) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(upload_events_every_ms_));
+
+      const std::vector<unsigned char>& upload_data_buffer = collector_.GetUploadDataBuffer();
+      total_uploaded_data_bytes_ += upload_data_buffer.size();
+      collector_.RefreshUploadDataBuffer();
+    }
+
+    absl::MutexLock lock{&upload_finished_mutex_};
+    upload_finished_ = true;
+  }
+
+  absl::Mutex stop_called_mutex_;
+  bool stop_called_ ABSL_GUARDED_BY(stop_called_mutex_) = false;
+
+  size_t event_count_ = 0;
+  size_t early_stop_event_count_;
+  uint64_t produce_event_every_ms_;
+  uint64_t upload_events_every_ms_;
+  std::thread data_upload_thread_;
+  std::thread data_produce_thread_;
+};
+
+}  // namespace
+
+TEST_F(UploaderClientCaptureEventCollectorTest, EarlyStop) {
+  constexpr size_t kEventCount = 10;
+  constexpr uint64_t kProduceEventEveryMs = 50;
+  constexpr uint64_t kUploadEventsEveryMs = 5;
+  constexpr size_t kEarlyStopEventCount = 4;
+  Start(kEventCount, kProduceEventEveryMs, kUploadEventsEveryMs, kEarlyStopEventCount);
+
+  // Stop producing data early
+  {
+    absl::MutexLock lock{&early_stop_mutex_};
+    early_stop_mutex_.Await(absl::Condition(&early_stop_));
```

Comment:
If I get this right you want to call `Stop` after some event have been processed and BEFORE the ProduceCaptureData thread has finished, right?

You will achieve the former with this construct but the latter is not guaranteed because the ProduceCaptureData thread can finish before this thread gets scheduled again.

To solve this you need to introduce another variable and mutex. I suggest to introduce an `int` counter which contains the number of events the ProduceCaptureData thread is allowed to send out. When it has reached the number it has to `Await` for the number to become higher. And the test needs to set it to an appropriate number when it wants the thread to continue.

```
// This needs to happen before the Producer thread is started
{
  absl::MutexLock lock{&number_of_allowed_events_mutex_};  
  number_of_allowed_events_ = 2;
}

// ...

{
    absl::MutexLock lock{&early_stop_mutex_};
    early_stop_mutex_.Await(absl::Condition(&early_stop_));
}

Stop();

// Stop was called, so now we can allow the Producer thread to continue
{
  absl::MutexLock lock{&number_of_allowed_events_mutex_};
  number_of_allowed_events_ = 4;
}

{
  absl::MutexLock lock{&upload_finished_mutex_};
  upload_finished_mutex_.Await(absl::Condition(&upload_finished_));
}
...
```

Reply:
Update: Obsolete as we removed the test for `EarlyStop`